src/coreclr/jit/gentree.cpp

```diff
@@ -25417,54 +26016,82 @@ GenTree* Compiler::gtNewSimdShuffleNode(
     var_types simdBaseType = JitType2PreciseVarType(simdBaseJitType);
     assert(varTypeIsArithmetic(simdBaseType));
 
-    if (op2->IsVectorAllBitsSet())
+    size_t elementSize  = genTypeSize(simdBaseType);
+    size_t elementCount = simdSize / elementSize;
+
+    // For ShuffleNative, delegate to the variable implementation to get the same behaviour for
+    // ShuffleNative with constant vs variable indices for free.
+    // We want ShuffleNative to be at least as good as Shuffle (at least in non out-of-range cases),
+    // so if we have all values in range, then just treat it like Shuffle.
+    // We may as well also track whether we have the identity shuffle and all out-of-range here.
+    bool gotInvalidIndex = false;
+    bool hasIdentityShuffle = true;
+    bool allOutOfRange = true;
+    for (size_t index = 0; index < elementCount; index++)
+    {
+        uint64_t value = op2->GetIntegralVectorConstElement(index, simdBaseType);
+        if (value >= elementCount)
+        {
+            gotInvalidIndex = true;
+        }
+        else
+        {
+            allOutOfRange = false;
+        }
+        if (value != static_cast<uint64_t>(index))
+        {
+            hasIdentityShuffle = false;
+        }
+    }
+    if (isShuffleNative && gotInvalidIndex)
+    {
+        // Call variable implementation.
+        return gtNewSimdShuffleNodeVariable(type, op1, op2, simdBaseJitType, simdSize, isShuffleNative);
+    }
+    if (hasIdentityShuffle)
     {
-        // AllBitsSet represents indices that are always "out of range" which means zero should be
+        // We have 0 1 2 ... for our indices, so just return op1
+        return op1;
+    }
+    if (allOutOfRange)
+    {
+        // allOutOfRange represents indices that are always "out of range" which means zero should be
         // selected for every element. We can special-case this down to just returning a zero node
-        return gtNewZeroConNode(type);
```

Comment:
@tannergooding for when you're reviewing: I believe the code here before was incorrect, as it didn't keep the side effects of `op1`. Consider something like this in the tree `V128.Create(/*something with side effects*/, V128<byte>.Zero)`. Please let me know if I'm missing something and should revert it.